Supporting Table Changes

zacdav-db opened the pull request:
Plan is to add/resolve the following:

- [x] Support loading table changes efficiently
  - [x] Expose table changes through existing `load_as_<format>` api via parameter `changes`
  - [x] Store table changes in `_table_changes` directory
  - [x] Store table in `_table` directory rather than root
  - [x] Add additional CDF columns efficiently
    - [x] Remove code that mutated files after download for CDF
    - [x] Transition to {duckdb} **_instead_** of {arrow} for CDF _only_ (review at later time for non-CDF)
      - [x] Write query/code to surface CDF table as expected
    - [x] Incorporate needed metadata into filenames
- [ ] Leverage existing cached data where possible (copy between `_table` and `_table_changes` directories)
- [x] Determine best method to manage duckdb connection
- [x] Add `list_all_tables()` method to sharing client
- [x] Adjust how version metadata is cleaned and presented
- [x] Add minimal example of CDF to README
- [x] Adjust README for breaking changes
- [x] Remove redundant validation logic
- [ ] Finalise TODO doc comments
- [x] Surface the data path (download path + table ID)
- [x] Add support for timestamp (0.6.0 protocol addition)
- [x] Increment version number
- [x] Change load functions (breaking)
- [x] Pass R CMD Check
- [x] Hide load_arrow_batch for now

tyler-richardett committed:
First pass at CDF functions

tyler-richardett committed:
Add CDF options error handling; refactor duplicative code; create separate subdirectory for table changes

zacdav-db committed:
Merge pull request #4 from tyler-richardett/table-changes

Supporting Table Changes

zacdav-db committed:
Changes:
 - remove renv files
 - added `list_all_tables()` to client
 - added retries to table$metadata
 - cleaned up table$current_version metadata
 - cleaned up loading functions - `load_as_<format>()` is always entry point (cdf and standard loading same func)
 - table changes are loaded into `_table_changes` rather than "table_changes"

zacdav-db committed:
Adjust README to have updated logic for CDF behaviour

zacdav-db commented:
To address "_Add additional CDF columns efficiently_" it is quite difficult without arrow's support for a function that is similar to `input_file_name()` which would let us avoid opening and mutating the parquets for add/remove file actions.

Currently I believe is unavoidable without doing an arrow based union which would be a compute penalty for every read rather than a one off cost at download.

zacdav-db commented:
> Currently I believe is unavoidable without doing an arrow based union which would be a compute penalty for every read rather than a one off cost at download.

Maybe I spoke too soon!
If we replace arrow specific read in code with duckdb (for at least CDF) then we can easily surface the additional CDF data on the fly through parsing filenames. This will enable the opening/closing of files to append additional columns for add/remove files.

```r
library(delta.sharing)
library(tidyverse)
library(duckdb)
library(DBI)

# create delta sharing client
client <- delta.sharing::sharing_client(credentials = "~/Desktop/config.share")

# access changes for a given table (10 most recent versions)
table <- client$table(share = "deltasharingr", schema = "simple", table = "cdf_no_partition")
table$set_download_path(path = "~/Desktop/share-download")
table$set_cdf_options(starting_version = table$current_version$version - 10)
tbl_cdf <- table$load_as_arrow(changes = TRUE)

# create duckdb connection
con = DBI::dbConnect(duckdb::duckdb(), dbdir = ":memory:")

# associate arrow dataset that was loaded with duckdb connection
# note: this in itself doesn't help much going from arrow --> duckdb, cool tho
duckdb::duckdb_register_arrow(conn = con, name = "test_cdf", arrow_scannable = tbl_cdf)
DBI::dbGetQuery(con, "SELECT *, input_file_name() FROM test_cdf LIMIT 10")

# lets manually try use duckdb to:
#   1. read data with file name
#   2. point to query upon data as arrow dataset
tbl_path <- file.path(table$path, table$metadata$metaData$id, "_table_changes/*")
query <- glue::glue_sql(
  "
  create table test_cdf2 as 
  select filename
  from read_parquet({tbl_path}, filename=true)
  ",
  .con = con,
)

# creates table
DBI::dbSendQuery(con, query)

# it works, we can see the filenames!
DBI::dbGetQuery(con, "select * from test_cdf2 limit 5")

# can use to_arrow() but it is not a Datase but a RecordBatchReader
test <- dplyr::tbl(con, "test_cdf2") %>% arrow::to_arrow()

# can still work with it though...
test %>%
  head()

# probably just expose as either:
# 1. dbplyr tbl
res_dbplyr <- dplyr::tbl(con, "test_cdf2") 
res_dbplyr %>%
  head() %>%
  mutate(x = 1)

# 2. tibble
res_dbplyr %>%
  collect()
```

zacdav-db committed:
Changes:
 - updated authors
 - added new helper `resolve_query_files()`
 - removed code to modify files on after download for CDF
 - removed helper `validate_starting_param`
 - adjusted `version_is_valid()`
 - adjusted `timestamp_is_valid()`
 - refactored `validate_cdf_options()`
 - added support for 0.6.0 protocol setting timestamp via `set_timestamp()`
 - WIP changes on `download()` to support copy of files to avoid re-download
 - added `table_path()` to extract root dir for data

zacdav-db committed:
Changes:
 - increment version number
 - changed read in to be via duckdb
 - adjust load functions

tyler-richardett commented on DESCRIPTION:
```diff
@@ -20,7 +24,9 @@ Encoding: UTF-8
 Imports:
   arrow,
+  DBI,
   dplyr,
```
Add `dbplyr`

zacdav-db replied:
It _should_ be okay with just dbplyr + duckdb + DBI.
RMD check indicates it's okay so far.

tyler-richardett replied:
Oh, weird. I installed all these dependencies, loaded the package, and got an error saying `dbplyr` still needed to be installed. (I would have assumed it to be a dependency of `duckdb` and/or `DBI`.) Perhaps an isolated issue.

tyler-richardett commented on R/helpers.R:
```diff
@@ -196,3 +196,248 @@ list_to_hive_partition <- function(x) {
+    )
+  }
+
+  if (all(!is.null(c(ending_version, ending_timestamp)))) {
```
`is.null` only returns one logical value (rather than a vector of equal length), so something like `ds_tbl$set_cdf_options(starting_version = 8, ending_version = 9)` now returns an error. Probably need to throw an `*apply` in there.

tyler-richardett commented on README.md:
```diff
@@ -36,4 +37,13 @@ ds_tbl_arrow %>%
+ds_tbl_cdf$set_cdf_options(starting_version = current_version - 10L)
+ds_tbl_cdf$starting_version
+
+ds_tbl_cdf_arrow <- ds_tbl_cdf$load_as_arrow(changes = TRUE)
```
`load_as_arrow` --> `load_arrow_table`

tyler-richardett commented on README.md:
```diff
@@ -36,4 +37,13 @@ ds_tbl_arrow %>%
+ds_tbl_cdf$starting_version
+
+ds_tbl_cdf_arrow <- ds_tbl_cdf$load_as_arrow(changes = TRUE)
+ds_tbl_cdf_tibble <- ds_tbl_cdf$load_as_tibble(changes = TRUE)
```
`load_as_tibble` --> `load_tibble`

zacdav-db committed:
Changes:
 - Hide loading as arrow batch functions
 - duckdb connection is now managed via R6 private field and passed from client

zacdav-db commented:
DuckDB connection is now created when the client is instantiated - this is then passed down to table class so that it can be -re-used.

May need to re-visit how disconnections are handled.

Commit:
improvements to stability of code
